third_party/move/move-vm/types/src/values/values_impl.rs

```diff
@@ -243,23 +243,69 @@ pub struct VectorRef(ContainerRef);
 enum GlobalValueImpl {
     /// No resource resides in this slot or in storage.
     None,
-    /// A resource has been published to this slot and it did not previously exist in storage.
-    Fresh { fields: Rc<RefCell<Vec<ValueImpl>>> },
+    /// A resource has been published to this slot and it did not previously exist in storage. The
+    /// invariant is that the value is a struct.
+    Fresh { value: Value },
     /// A resource resides in this slot and also in storage. The status flag indicates whether
     /// it has potentially been altered.
     Cached {
-        fields: Rc<RefCell<Vec<ValueImpl>>>,
+        /// A struct value representing this resource (invariant).
+        value: Value,
         status: Rc<RefCell<GlobalDataStatus>>,
     },
     /// A resource used to exist in storage but has been deleted by the current transaction.
     Deleted,
 }
 
-/// A wrapper around `GlobalValueImpl`, representing a "slot" in global storage that can
-/// hold a resource.
+/// Represents a "slot" in global storage that can hold a resource. The resource is always a struct
+/// or an enum.
+///
+/// IMPORTANT: [Clone] is not implemented for a reason, use [Copyable] trait for deep copies.
 #[derive(Debug)]
 pub struct GlobalValue(GlobalValueImpl);
 
+/// Trait to represent copyable values so that [GlobalValue] can support copy-on-write. Note that
+/// we explicitly do not implement [Clone].
+pub trait Copyable: Sized {
+    fn deep_copy(&self) -> PartialVMResult<Self>;
+}
+
+impl<T> Copyable for T
+where
+    T: Clone,
+{
+    fn deep_copy(&self) -> PartialVMResult<Self> {
+        Ok(self.clone())
+    }
+}
+
+impl Copyable for GlobalValue {
+    fn deep_copy(&self) -> PartialVMResult<Self> {
+        Ok(Self(match &self.0 {
+            GlobalValueImpl::None => GlobalValueImpl::None,
+            GlobalValueImpl::Fresh { value } => {
+                let value = value
+                    .0
+                    .copy_value(1, Some(DEFAULT_MAX_VM_VALUE_NESTED_DEPTH))?;
+                GlobalValueImpl::Fresh {
+                    value: Value(value),
+                }
+            },
+            GlobalValueImpl::Cached { value, status } => {
+                let value = value
+                    .0
+                    .copy_value(1, Some(DEFAULT_MAX_VM_VALUE_NESTED_DEPTH))?;
```

Comment:
are these supposed to fail or it's invariant?

Reply:
Copying the value should never fail, as depth is the same as when it was constructed. I kept the same flow as before to avoid expects in case things change, but we can also map this to invariant violation, not sure if it makes a difference?